### ISSUE TITLE
sql: Composite Foreign Key Matching

### ISSUE THREAD

**Issue by BramGruneir** [cockroachdb/cockroach#20305]
Here is a rundown of the three type of composite foreign key matching:

* `MATCH FULL` considers a foreign key as valid if every foreign key column is
  null or if all foreign key columns are not null and match the referenced
  table.
* `MATCH PARTIAL` is valid if at least one Foreign key column is null and the
  rest of the not null columns match those of the referenced tables or all
  foreign key columns are not null and match the referenced table.
* `MATCH SIMPLE` considers a foreign key valid if all columns are not null and
  match the referenced key exactly or if one of the foreign key values is null.

Currently, we only offer `MATCH FULL`.
Postgres only offer `MATCH FULL` and `MATCH SIMPLE` with `MATCH SIMPLE` as the default.

We should add in `MATCH SIMPLE` and set it as our default for all new composite foreign keys references.  This may require a migration of all pre-existing composite foreign key references to `MATCH SIMPLE` or explicitly set them `MATCH FULL`.

**Comment by vivekmenezes** [comment]
Why does this need a migration? We currently default to MATCH FULL. We can change that default to MATCH SIMPLE. Will that affect anyone's existing data?

**Comment by BramGruneir** [comment]
It might.  It could mean the difference between allowing a delete/update operation or not.

Changing the matching scheme out from under someone seems like a pretty poor thing to do.

**Comment by vivekmenezes** [comment]
It's pretty subtle but we're committing to postgres compatibility as much as possible so now's a good time to fix this. Looks like MATCH_SIMPLE is going to allow more matches. Any strong reason to not fix this now?

**Comment by BramGruneir** [comment]
Just that for everything we do change, something else gets pushed.  But I'll leave the planning and prioritizing to the PMs.  If anything, I think it should wait until the cascading features are complete.

**Comment by BramGruneir** [comment]
@dt Any comments on this?  I know you looked into it before.

**Comment by BramGruneir** [comment]
I do think this would be a good starter project to be honest.

**Comment by dt** [comment]
I didn't look into this much -- you probably have more context than I do at this point. The difference between SIMPLE and PARTIAL seem pretty small?

**Comment by BramGruneir** [comment]
Indeed.  However, partial is NOT supported by postgres yet.  But I don't see why we would handle all 3.

From https://www.postgresql.org/docs/current/static/sql-createtable.html
```
A value inserted into the referencing column(s) is matched against the values of the referenced table and referenced columns using the given match type. There are three match types: MATCH FULL, MATCH PARTIAL, and MATCH SIMPLE (which is the default). MATCH FULL will not allow one column of a multicolumn foreign key to be null unless all foreign key columns are null; if they are all null, the row is not required to have a match in the referenced table. MATCH SIMPLE allows any of the foreign key columns to be null; if any of them are null, the row is not required to have a match in the referenced table. MATCH PARTIAL is not yet implemented. (Of course, NOT NULL constraints can be applied to the referencing column(s) to prevent these cases from arising.)
```

**Comment by dt** [comment]
so, SIMPLE means if anything is null, don't check the rest? that's... ok. That certainly isn't what I'd pick for a default, but, I've learned before to just do what postgres does and not argue.

**Comment by vivekmenezes** [comment]
Yes this issue is about doing what postgres does unless there is a strong
argument not to

**Comment by BramGruneir** [comment]
We need to resurrect this issue.

In light of #30026, we I feel there is really only a few correct actions here:

1) We should move to support MATCH SIMPLE by default and convert all current FK relationships to that.  We cannot move them to MATCH FULL as any mixed NULL/NON NULL will suddenly not work.  But MATCH SIMPLE would work. This needs to be communicated as a breaking change in behaviour.
2) We should add support for MATCH FULL, correctly, this should not be too difficult, since we just need to add guardrails when inserting and the current logic is mostly correct.
3) We should force all new INTERLEAVED (and maybe partitioning) matching to use it. This would prevent interleaved tables from being able to be orphaned from its parents.  This along with #28537 would go a long to strengthening interleaved tables.

As much I like the idea of match partial, there is potential for a fairly large blow up of the search space when it's used.  So we should leave it for now as unimplemented until there is a serious reason to do so.  Postgres doesn't implement it and this seems like a good choice.

**Comment by BramGruneir** [comment]
Update the original comment here to make the descriptions clearer.  Match full isn't supported yet, but will be as soon as #32998 lands.
